crates/neon/src/types_impl/map.rs

```diff
@@ -0,0 +1,163 @@
+use std::{error, fmt, mem::MaybeUninit};
+
+use crate::{
+    context::{internal::Env, Context},
+    handle::{internal::TransparentNoCopyWrapper, root::NapiRef, Handle},
+    macro_internal::NeonResultTag,
+    object::Object,
+    result::{JsResult, NeonResult, ResultExt},
+    sys::{self, raw},
+    types::{private, JsFunction, JsObject, Value},
+};
+
+use super::{JsBoolean, JsNumber, JsUndefined};
+
+#[derive(Debug)]
+#[repr(transparent)]
+/// The type of JavaScript
+/// [`Map`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map)
+/// objects.
+pub struct JsMap(raw::Local);
+
+impl JsMap {
+    pub fn new<'cx, C>(cx: &mut C) -> NeonResult<Handle<'cx, Self>>
+    where
+        C: Context<'cx>,
+    {
+        let map = cx
+            .global::<JsFunction>("Map")?
+            .construct_with(cx)
+            .apply::<JsObject, _>(cx)?;
+
+        Ok(map.downcast_or_throw(cx)?)
+    }
+
+    pub fn size<'a, C: Context<'a>>(&self, cx: &mut C) -> NeonResult<Handle<'a, JsNumber>> {
+        Object::get(self, cx, "size")
+    }
+
+    // TODO: is the return type important here ?
+    // I see three possibilities here:
+    // 1. Stick to the JS and return the `undefined` (this is what we do now)
+    // 2. Check we get an `undefined` and return `Ok(())`
+    // 3. Just discard the return value and return `Ok(())`
+    // Solutions 2 & 3 are more user-friendly, but make more assumptions (though it
+    // should be fine given `Map` is not expected to be overridden ?).
+    pub fn clear<'a, C: Context<'a>>(&self, cx: &mut C) -> NeonResult<Handle<'a, JsUndefined>> {
+        Object::call_method_with(self, cx, "clear")?.apply(cx)
+    }
+
+    pub fn delete<'a, C: Context<'a>, K: Value>(
```

Comment:
If we are going to have wrappers for the built in methods, we should try to make them feel more Rust-y. Currently, this requires calling like:

```rust
ket k = cx.string("key");
map.delete(&mut cx, k)?;
```

Ideally, I could call:

```rust
map.delete(&mut cx, "key")?;
```

My suggestion is to accept `K: TryIntoJs<'cx>` instead of a handle. Additionally, with the addition of `Cx` there isn't a compelling reason to be generic over `Context`.

```rust
impl JsMap {
    pub fn delete<'cx, K>(cx: &mut Cx<'cx>, k: K) -> NeonResult<bool>
    where
        K: TryIntoJs<'cx>,
    {
        self.prop(&mut cx, "delete").bind()?.arg(k)?.call()
    }
}
```

Reply:
So you mean `Cx` is the new way of doing things and `C: Context<'a>` the old one ?
I was wondering why there was those two and wasn't sure if `C: Context<'a>` wasn't present for a specific need regarding genericity...

Reply:
`Context` is the old way and `Cx` is the new way. `Cx` was added to support places where we *can't* be generic because of limitations of GAT (e.g., `Fn*` traits).

Sometimes it's still useful to be generic over `Context` if you hand the `Context` back to user controlled code because `FunctionContext` has additional capabilities. However, if the implementation is fully generic over `Context`, `Cx` makes for a much simpler function signature.

From a developer experience perspective, they are equivalent because all contexts implement `RefMut<Cx>`.